### ISSUE TITLE
Allowing conditions to be used instead of just integers for... everything

### ISSUE THREAD

**Issue by Hurleveur** [endless-sky/endless-sky#7861]
### Problem Description

Currently we cannot easily make a mission adapt to the strength of the player, let's take an example: say I want the Unfettered to fight the player with just what is needed for it to be a fair fight (or to beat the player caus he's been nasty), the only way to do it is by using multiple npcs that each have a 
to spawn
"condition" > X

Another use case is removing all outfits of one kind you have. Or simply making a mission that takes all the bunks you have available. Possibilities are endless.
Of course, the number would not stay in flux, it'd be written down as an integer as soon as the mission is loaded, so that state is what will matter. (this way is much more elegant and simpler)

### Related Issue Links

I looked but could not find any, but I'm not very experienced at finding these things so maybe I missed it.

### Desired Solution

Being able to use any condition wherever a hard set number can be given.
For instance, in a mission, one could write:
mission "take ur space"
cargo "some cargo" <"ANY CONDITION"/X>

When the mission is created it will look what the condition is (they can only be integers anyway) and replace it.

another example, where one would set power before (in the offering of the mission of instance), we could have that amount of shield beetles spawning, like this:
npc
ship "Shield Beetle" "power"

Yet another example, for outfits being given/taken away one could write:
outfit "Jump Drive" -"outfit: Jump Drive"
which would remove all jump drives. I want this kind of thing to remove all gold bars too, in the Unfettered campaign, and before that give you a payment corresponding to that amount.

### Alternative Approaches

I don't really see any.

### Additional Context

n/a

**Comment by UnorderedSigh** [comment]
When should the variables for the npc be expanded?

**Comment by Hurleveur** [comment]
No I just will make the missions trigger at the last time and I'll get what's needed. We can always expand it later but I feel like its best to keep it like this for now, that way almost nothing changes.

**Comment by UnorderedSigh** [comment]
Are you saying you want the variables to be expanded when the mission triggers?

That means the mission would have to store the variable name for each quantity when it is Load()ed from game files.

**Comment by UnorderedSigh** [comment]
Actually, it's more than just the Mission. The NPC, Fleet, Ship, and quite a lot of other things would have to store that information.

**Comment by Hurleveur** [comment]
I'm saying it should not store the variable for now and just write the number that corresponds to the state of the variable at the time of mission creation.

**Comment by UnorderedSigh** [comment]
The Mission object is created when the game data is loaded, so the value you'll get is whatever was set in the save file when the game data was loaded.

**Comment by UnorderedSigh** [comment]
> The Mission object is created when the game data is loaded, so the value you'll get is whatever was set in the save file when the game data was loaded.

I'm not sure if that's right. Once the medicine the doctor pumped into me wears off, I'll check this again.

**Comment by Hurleveur** [comment]
No, the mission object is pre-loaded whenever the mission can be offered.

**Comment by UnorderedSigh** [comment]
Then I think we need a `DataNode::ValueLookup(int index,const ConditionsStore &)` and somehow the Load() routines would need the ConditionsStore.

**Comment by UnorderedSigh** [comment]
Unfortunately, you are incorrect. The missions are loaded once, and put in GameData.missions. The PlayerInfo::CreateMissions calls GameData::Missions() to get the list, and then calls Instantiate on it. That means Load() is called only once for each Mission.

To implement what you suggest, Mission and all the classes it uses, will have to keep the variable or integer for each thing it reads in. Then, when Instantiate is called, it'll generate the integer. It isn't as simple as throwing an integer around, though. In many cases, Load() will not store the integer at all; it uses the data immediately and discards the integer.

**Comment by UnorderedSigh** [comment]
The only way to do this is one step at a time. Make a CalculatedNumber class (or something like that) and put it into the various parts of the Mission object tree, one-by-one.

**Comment by UnorderedSigh** [comment]
I'm going to try this with a tiny part of the Mission object tree and do a PR to see how people like it. Then we can move on to the rest of the Mission object tree.

**Comment by Hurleveur** [comment]
I mean instantiate already has access to the player doesnt it?
But yeah I meant from a player perspective
May as well make the mission replace all integers by a class that will give its value once its given a ref to the player's conditions?

**Comment by UnorderedSigh** [comment]
The Instantiate() functions don't receive the PlayerInfo, but could be added easily since that information is available when Mission::Instantiate() is called. I am working on a prototype of this, which will replace some of the integers with objects. Then people can look at it and decide if it is the right way.

**Comment by petervdmeer** [comment]
I created an example of how to use this feature for variable outfit costs with the recently updated ConditionSet code; https://github.com/endless-sky/endless-sky/pull/11289 . It is not fully working yet, but I guess the idea should already be clear.
